File: ui/portal/src/types/submission.ts
/**
 * Submission tracking types for the ATS functionality.
 * Handles candidate submissions to job postings.
 */

// ==================== Constants ====================

export const SUBMISSION_STATUSES = [
  'SUBMITTED',
  'CLIENT_REVIEW',
  'INTERVIEW_SCHEDULED',
  'INTERVIEWED',
  'OFFERED',
  'PLACED',
  'REJECTED',
  'WITHDRAWN',
  'ON_HOLD',
] as const;

export type SubmissionStatus = (typeof SUBMISSION_STATUSES)[number];

export const RATE_TYPES = ['HOURLY', 'DAILY', 'WEEKLY', 'MONTHLY', 'ANNUAL'] as const;
export type RateType = (typeof RATE_TYPES)[number];

export const INTERVIEW_TYPES = ['PHONE', 'VIDEO', 'ONSITE', 'TECHNICAL', 'HR', 'PANEL'] as const;
export type InterviewType = (typeof INTERVIEW_TYPES)[number];

export const PRIORITY_LEVELS = ['HIGH', 'MEDIUM', 'LOW'] as const;
export type PriorityLevel = (typeof PRIORITY_LEVELS)[number];

export const ACTIVITY_TYPES = [
  'CREATED',
  'STATUS_CHANGE',
  'NOTE',
  'EMAIL_SENT',
  'EMAIL_RECEIVED',
  'CALL_LOGGED',
  'INTERVIEW_SCHEDULED',
  'INTERVIEW_COMPLETED',
  'INTERVIEW_CANCELLED',
  'RATE_UPDATED',
  'VENDOR_UPDATED',
  'PRIORITY_CHANGED',
  'FOLLOW_UP_SET',
  'RESUME_SENT',
  'CLIENT_FEEDBACK',
] as const;

export type ActivityType = (typeof ACTIVITY_TYPES)[number];

// Status display labels and colors
export const STATUS_LABELS: Record<SubmissionStatus, string> = {
  SUBMITTED: 'Submitted',
  CLIENT_REVIEW: 'Client Review',
  INTERVIEW_SCHEDULED: 'Interview Scheduled',
  INTERVIEWED: 'Interviewed',
  OFFERED: 'Offered',
  PLACED: 'Placed',
  REJECTED: 'Rejected',
  WITHDRAWN: 'Withdrawn',
  ON_HOLD: 'On Hold',
};

export const STATUS_COLORS: Record<SubmissionStatus, string> = {
  SUBMITTED: 'bg-blue-100 text-blue-800',
  CLIENT_REVIEW: 'bg-yellow-100 text-yellow-800',
  INTERVIEW_SCHEDULED: 'bg-purple-100 text-purple-800',
  INTERVIEWED: 'bg-indigo-100 text-indigo-800',
  OFFERED: 'bg-emerald-100 text-emerald-800',
  PLACED: 'bg-green-100 text-green-800',
  REJECTED: 'bg-red-100 text-red-800',
  WITHDRAWN: 'bg-gray-100 text-gray-800',
  ON_HOLD: 'bg-orange-100 text-orange-800',
};

export const PRIORITY_COLORS: Record<PriorityLevel, string> = {
  HIGH: 'bg-red-100 text-red-800',
  MEDIUM: 'bg-yellow-100 text-yellow-800',
  LOW: 'bg-gray-100 text-gray-800',
};

// Terminal statuses (no further transitions)
export const TERMINAL_STATUSES: SubmissionStatus[] = ['PLACED', 'REJECTED', 'WITHDRAWN'];

// Active statuses (in progress)
export const ACTIVE_STATUSES: SubmissionStatus[] = [
  'SUBMITTED',
  'CLIENT_REVIEW',
  'INTERVIEW_SCHEDULED',
  'INTERVIEWED',
  'OFFERED',
  'ON_HOLD',
];

// ==================== Nested Response Types ====================

export interface SubmissionCandidate {
  id: number;
  first_name: string;
  last_name?: string;
  email?: string;
  phone?: string;
  current_title?: string;
  skills?: string[];
  location?: string;
  visa_type?: string; // Visa/work authorization type
}

export interface SubmissionJob {
  id?: number | null; // Null for external jobs
  title: string;
  company: string;
  location?: string;
  job_type?: string;
  is_remote?: boolean;
  platform?: string; // 'external' for external jobs
  job_url?: string;
}

export interface SubmissionUser {
  id: number;
  email: string;
  first_name?: string;
  last_name?: string;
}

// ==================== Activity Types ====================

export interface SubmissionActivity {
  id: number;
  submission_id: number;
  activity_type: ActivityType;
  content?: string;
  old_value?: string;
  new_value?: string;
  metadata?: Record<string, unknown>;
  created_at: string;
  created_by_id?: number;
  created_by?: SubmissionUser;
}

// ==================== Main Submission Type ====================

export interface Submission {
  id: number;
  candidate_id: number;
  job_posting_id?: number | null; // Nullable for external jobs
  submitted_by_user_id?: number;
  tenant_id: number;

  // External Job Info
  is_external_job?: boolean;
  external_job_title?: string;
  external_job_company?: string;
  external_job_location?: string;
  external_job_url?: string;
  external_job_description?: string;

  // Status
  status: SubmissionStatus;
  status_changed_at?: string;

  // Vendor info
  vendor_company?: string;
  vendor_contact_name?: string;
  vendor_contact_email?: string;
  vendor_contact_phone?: string;
  client_company?: string;

  // Rates
  bill_rate?: number;
  pay_rate?: number;
  rate_type?: RateType;
  currency?: string;
  margin?: number;
  margin_percentage?: number;

  // Notes
  submission_notes?: string;
  cover_letter?: string;
  tailored_resume_id?: number;

  // Interview
  interview_scheduled_at?: string;
  interview_type?: InterviewType;
  interview_location?: string;
  interview_notes?: string;
  interview_feedback?: string;

  // Outcome
  rejection_reason?: string;
  rejection_stage?: string;
  withdrawal_reason?: string;

  // Placement
  placement_start_date?: string;
  placement_end_date?: string;
  placement_duration_months?: number;

  // Priority
  priority?: PriorityLevel;
  is_hot?: boolean;
  follow_up_date?: string;

  // Timestamps
  submitted_at: string;
  created_at: string;
  updated_at: string;

  // Computed fields
  days_since_submitted?: number;
  is_active?: boolean;

  // Nested objects (optional, included based on endpoint)
  candidate?: SubmissionCandidate;
  job?: SubmissionJob;
  submitted_by?: SubmissionUser;
  activities?: SubmissionActivity[];
}

// ==================== Request Types ====================

export interface SubmissionCreateInput {
  candidate_id: number;
  job_posting_id: number;

  vendor_company: string;
  vendor_contact_name: string;
  vendor_contact_email: string;
  vendor_contact_phone: string;
  client_company: string;

  bill_rate: number;
  pay_rate: number;
  rate_type?: RateType;
  currency?: string;

  submission_notes: string;
  cover_letter?: string;
  tailored_resume_id?: number;

  priority?: PriorityLevel;
  is_hot?: boolean;
  follow_up_date?: string;
}

export interface ExternalSubmissionCreateInput {
  // Required
  candidate_id: number;

  // External job info (required)
  external_job_title: string;
  external_job_company: string;
  external_job_location?: string;
  external_job_url?: string;
  external_job_description?: string;

  // Vendor/Client info
  vendor_company?: string;
  vendor_contact_name?: string;
  vendor_contact_email?: string;
  vendor_contact_phone?: string;
  client_company?: string;

  // Rate information
  bill_rate?: number;
  pay_rate?: number;
  rate_type?: RateType;
  currency?: string;

  // Submission details
  submission_notes?: string;

  // Priority
  priority?: PriorityLevel;
  is_hot?: boolean;
  follow_up_date?: string;
}

export interface SubmissionUpdateInput {
  // Vendor/Client info
  vendor_company?: string;
  vendor_contact_name?: string;
  vendor_contact_email?: string;
  vendor_contact_phone?: string;
  client_company?: string;

  // Rate information
  bill_rate?: number;
  pay_rate?: number;
  rate_type?: RateType;
  currency?: string;

  // Submission details
  submission_notes?: string;
  cover_letter?: string;
  tailored_resume_id?: number;

  // Interview info
  interview_scheduled_at?: string;
  interview_type?: InterviewType;
  interview_location?: string;
  interview_notes?: string;
  interview_feedback?: string;

  // Priority
  priority?: PriorityLevel;
  is_hot?: boolean;
  follow_up_date?: string;
}

export interface SubmissionStatusUpdateInput {
  status: SubmissionStatus;
  note?: string;

  // Status-specific fields
  rejection_reason?: string;
  rejection_stage?: string;
  withdrawal_reason?: string;

  // For placement status
  placement_start_date?: string;
  placement_end_date?: string;
  placement_duration_months?: number;
}

export interface SubmissionInterviewScheduleInput {
  interview_scheduled_at: string;
  interview_type: InterviewType;
  interview_location?: string;
  interview_notes?: string;
}

export interface SubmissionActivityCreateInput {
  activity_type?: ActivityType;
  content: string;
  metadata?: Record<string, unknown>;
}

// ==================== Filter Types ====================

export interface SubmissionFilters {
  status?: SubmissionStatus;
  statuses?: SubmissionStatus[];
  candidate_id?: number;
  job_posting_id?: number;
  submitted_by_user_id?: number;
  vendor_company?: string;
  client_company?: string;
  priority?: PriorityLevel;
  is_hot?: boolean;
  is_active?: boolean;

  // Date filters
  submitted_after?: string;
  submitted_before?: string;
  interview_after?: string;
  interview_before?: string;

  // Pagination
  page?: number;
  per_page?: number;

  // Sorting
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

// ==================== Response Types ====================

export interface SubmissionListResponse {
  items: Submission[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
}

export interface SubmissionStats {
  total: number;
  by_status: Record<SubmissionStatus, number>;
  submitted_this_week: number;
  submitted_this_month: number;
  interviews_scheduled: number;
  placements_this_month: number;
  average_days_to_placement?: number;
  interview_rate?: number;
  placement_rate?: number;
}

export interface SubmissionFollowUp {
  id: number;
  candidate_name: string;
  job_title: string;
  company: string;
  follow_up_date: string;
  status: SubmissionStatus;
  is_overdue: boolean;
}

export interface SubmissionFollowUpsResponse {
  upcoming: SubmissionFollowUp[];
  overdue: SubmissionFollowUp[];
}

// ==================== Utility Types ====================

// For the submission dialog / quick submit
export interface QuickSubmitData {
  candidate_id: number;
  job_posting_id: number;
  candidate_name?: string;
  job_title?: string;
  company?: string;
}

// Status transition validation
export const VALID_STATUS_TRANSITIONS: Record<SubmissionStatus, SubmissionStatus[]> = {
  SUBMITTED: ['CLIENT_REVIEW', 'INTERVIEW_SCHEDULED', 'REJECTED', 'WITHDRAWN', 'ON_HOLD'],
  CLIENT_REVIEW: ['INTERVIEW_SCHEDULED', 'REJECTED', 'WITHDRAWN', 'ON_HOLD'],
  INTERVIEW_SCHEDULED: ['INTERVIEWED', 'REJECTED', 'WITHDRAWN', 'ON_HOLD'],
  INTERVIEWED: ['OFFERED', 'REJECTED', 'WITHDRAWN', 'ON_HOLD'],
  OFFERED: ['PLACED', 'REJECTED', 'WITHDRAWN', 'ON_HOLD'],
  PLACED: [], // Terminal
  REJECTED: [], // Terminal
  WITHDRAWN: [], // Terminal
  ON_HOLD: ['SUBMITTED', 'CLIENT_REVIEW', 'INTERVIEW_SCHEDULED', 'INTERVIEWED', 'OFFERED', 'WITHDRAWN'],
};

// Helper to check if a status transition is valid
export function isValidStatusTransition(
  currentStatus: SubmissionStatus,
  newStatus: SubmissionStatus
): boolean {
  return VALID_STATUS_TRANSITIONS[currentStatus]?.includes(newStatus) ?? false;
}

// Helper to get next valid statuses
export function getNextValidStatuses(currentStatus: SubmissionStatus): SubmissionStatus[] {
  return VALID_STATUS_TRANSITIONS[currentStatus] ?? [];
}

// Helper to check if status is terminal
export function isTerminalStatus(status: SubmissionStatus): boolean {
  return TERMINAL_STATUSES.includes(status);
}

// Helper to check if status is active
export function isActiveStatus(status: SubmissionStatus): boolean {
  return ACTIVE_STATUSES.includes(status);
}

File: server/app/models/submission.py
"""
Submission Model
Tracks candidate submissions to job postings by recruiters.
Core model for the ATS (Applicant Tracking System) functionality.
"""
from datetime import datetime
from decimal import Decimal
from sqlalchemy import (
    String, Integer, Text, DateTime, Boolean, 
    ForeignKey, Index, DECIMAL, ARRAY
)
from sqlalchemy.dialects.postgresql import JSONB
from app import db


class SubmissionStatus:
    """Submission status constants."""
    SUBMITTED = "SUBMITTED"           # Initial submission to vendor
    CLIENT_REVIEW = "CLIENT_REVIEW"   # Vendor sent to client
    INTERVIEW_SCHEDULED = "INTERVIEW_SCHEDULED"  # Interview set up
    INTERVIEWED = "INTERVIEWED"       # Interview completed
    OFFERED = "OFFERED"               # Offer extended
    PLACED = "PLACED"                 # Candidate placed (won!)
    REJECTED = "REJECTED"             # Rejected at any stage
    WITHDRAWN = "WITHDRAWN"           # Recruiter withdrew submission
    ON_HOLD = "ON_HOLD"               # Temporarily on hold
    
    @classmethod
    def all(cls) -> list[str]:
        return [
            cls.SUBMITTED, cls.CLIENT_REVIEW, cls.INTERVIEW_SCHEDULED,
            cls.INTERVIEWED, cls.OFFERED, cls.PLACED, 
            cls.REJECTED, cls.WITHDRAWN, cls.ON_HOLD
        ]
    
    @classmethod
    def active(cls) -> list[str]:
        """Statuses that are still in progress."""
        return [
            cls.SUBMITTED, cls.CLIENT_REVIEW, cls.INTERVIEW_SCHEDULED,
            cls.INTERVIEWED, cls.OFFERED, cls.ON_HOLD
        ]
    
    @classmethod
    def terminal(cls) -> list[str]:
        """Final statuses (no further action)."""
        return [cls.PLACED, cls.REJECTED, cls.WITHDRAWN]


class Submission(db.Model):
    """
    Tracks when a recruiter submits a candidate to a job posting.
    
    This is the core ATS model for bench recruiters to track:
    - Which candidates were submitted to which jobs
    - Current status in the hiring pipeline
    - Vendor/client information
    - Bill and pay rates
    - Activity history
    """
    __tablename__ = 'submissions'
    
    id = db.Column(Integer, primary_key=True)
    
    # Core Relationships
    candidate_id = db.Column(
        Integer, 
        ForeignKey('candidates.id', ondelete='CASCADE'), 
        nullable=False, 
        index=True
    )
    # job_posting_id is nullable to support external/manual job submissions
    job_posting_id = db.Column(
        Integer, 
        ForeignKey('job_postings.id', ondelete='CASCADE'), 
        nullable=True,  # Nullable for external jobs
        index=True
    )
    
    # External Job Fields (for jobs not in the portal)
    # These are used when job_posting_id is NULL
    is_external_job = db.Column(Boolean, default=False, nullable=False)
    external_job_title = db.Column(String(255))
    external_job_company = db.Column(String(255))
    external_job_location = db.Column(String(255))
    external_job_url = db.Column(String(1000))  # URL to the original job posting
    external_job_description = db.Column(Text)  # Brief description or notes about the job
    submitted_by_user_id = db.Column(
        Integer, 
        ForeignKey('portal_users.id', ondelete='SET NULL'), 
        nullable=True, 
        index=True
    )
    tenant_id = db.Column(
        Integer, 
        ForeignKey('tenants.id', ondelete='CASCADE'), 
        nullable=False, 
        index=True
    )
    
    # Status Tracking
    status = db.Column(
        String(50), 
        nullable=False, 
        default=SubmissionStatus.SUBMITTED,
        index=True
    )
    status_changed_at = db.Column(DateTime, default=datetime.utcnow)
    status_changed_by_id = db.Column(
        Integer, 
        ForeignKey('portal_users.id', ondelete='SET NULL'),
        nullable=True
    )
    
    # Vendor/Client Information
    vendor_company = db.Column(String(255))  # Vendor/Staffing agency name
    vendor_contact_name = db.Column(String(255))  # Contact person
    vendor_contact_email = db.Column(String(255))  # Contact email
    vendor_contact_phone = db.Column(String(50))  # Contact phone
    client_company = db.Column(String(255))  # End client company (if known)
    
    # Rate Information
    bill_rate = db.Column(DECIMAL(10, 2))  # What client pays ($/hr)
    pay_rate = db.Column(DECIMAL(10, 2))   # What candidate gets ($/hr)
    rate_type = db.Column(String(20), default='HOURLY')  # HOURLY, DAILY, WEEKLY, MONTHLY, ANNUAL
    currency = db.Column(String(10), default='USD')
    
    # Submission Details
    submission_notes = db.Column(Text)  # Notes about the submission
    cover_letter = db.Column(Text)  # Cover letter if sent
    resume_version_used = db.Column(String(255))  # Which resume was sent
    tailored_resume_id = db.Column(
        Integer, 
        ForeignKey('tailored_resumes.id', ondelete='SET NULL'),
        nullable=True
    )  # If a tailored resume was used
    
    # Interview Information
    interview_scheduled_at = db.Column(DateTime)  # Interview date/time
    interview_type = db.Column(String(50))  # PHONE, VIDEO, ONSITE, TECHNICAL
    interview_location = db.Column(String(500))  # Location or video link
    interview_notes = db.Column(Text)  # Pre-interview notes
    interview_feedback = db.Column(Text)  # Post-interview feedback
    
    # Outcome Information
    rejection_reason = db.Column(Text)  # If rejected, why
    rejection_stage = db.Column(String(50))  # At which stage rejected
    withdrawal_reason = db.Column(Text)  # If withdrawn, why
    
    # Placement Information (if placed)
    placement_start_date = db.Column(DateTime)  # When candidate starts
    placement_end_date = db.Column(DateTime)  # Contract end date (if known)
    placement_duration_months = db.Column(Integer)  # Expected duration
    
    # Priority and Flags
    priority = db.Column(String(20), default='MEDIUM')  # HIGH, MEDIUM, LOW
    is_hot = db.Column(Boolean, default=False)  # Hot/urgent submission
    follow_up_date = db.Column(DateTime)  # When to follow up
    
    # Timestamps
    submitted_at = db.Column(DateTime, default=datetime.utcnow, nullable=False)
    created_at = db.Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = db.Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    candidate = db.relationship('Candidate', backref=db.backref('submissions', lazy='dynamic'))
    job_posting = db.relationship('JobPosting', backref=db.backref('submissions', lazy='dynamic'))
    submitted_by = db.relationship(
        'PortalUser', 
        foreign_keys=[submitted_by_user_id],
        backref='submissions_made'
    )
    status_changed_by = db.relationship(
        'PortalUser',
        foreign_keys=[status_changed_by_id],
        backref='submission_status_changes'
    )
    tenant = db.relationship('Tenant', backref=db.backref('submissions', lazy='dynamic'))
    tailored_resume = db.relationship('TailoredResume', backref='submission')
    activities = db.relationship(
        'SubmissionActivity',
        back_populates='submission',
        cascade='all, delete-orphan',
        order_by='desc(SubmissionActivity.created_at)'
    )
    
    # Indexes
    __table_args__ = (
        # Unique constraint: one submission per candidate per job per tenant
        # Note: This only applies to internal jobs (job_posting_id NOT NULL)
        # External jobs are not constrained by this index
        Index('idx_submission_unique', 'candidate_id', 'job_posting_id', 'tenant_id', unique=True, postgresql_where=db.text('job_posting_id IS NOT NULL')),
        # Common query patterns
        Index('idx_submission_tenant_status', 'tenant_id', 'status'),
        Index('idx_submission_submitted_at', 'submitted_at', postgresql_ops={'submitted_at': 'DESC'}),
        Index('idx_submission_submitted_by', 'submitted_by_user_id', 'status'),
        # External job queries
        Index('idx_submission_external', 'tenant_id', 'is_external_job'),
    )
    
    def __repr__(self):
        return f'<Submission {self.id} candidate={self.candidate_id} job={self.job_posting_id} status={self.status}>'
    
    def to_dict(
        self, 
        include_candidate: bool = False, 
        include_job: bool = False,
        include_submitted_by: bool = False,
        include_activities: bool = False,
        include_rates: bool = True
    ) -> dict:
        """Convert submission to dictionary."""
        result = {
            'id': self.id,
            'candidate_id': self.candidate_id,
            'job_posting_id': self.job_posting_id,
            'submitted_by_user_id': self.submitted_by_user_id,
            'tenant_id': self.tenant_id,
            
            # External Job Info
            'is_external_job': self.is_external_job,
            'external_job_title': self.external_job_title,
            'external_job_company': self.external_job_company,
            'external_job_location': self.external_job_location,
            'external_job_url': self.external_job_url,
            'external_job_description': self.external_job_description,
            
            # Status
            'status': self.status,
            'status_changed_at': self.status_changed_at.isoformat() if self.status_changed_at else None,
            
            # Vendor Info
            'vendor_company': self.vendor_company,
            'vendor_contact_name': self.vendor_contact_name,
            'vendor_contact_email': self.vendor_contact_email,
            'vendor_contact_phone': self.vendor_contact_phone,
            'client_company': self.client_company,
            
            # Notes
            'submission_notes': self.submission_notes,
            'tailored_resume_id': self.tailored_resume_id,
            
            # Interview
            'interview_scheduled_at': self.interview_scheduled_at.isoformat() if self.interview_scheduled_at else None,
            'interview_type': self.interview_type,
            'interview_location': self.interview_location,
            'interview_notes': self.interview_notes,
            'interview_feedback': self.interview_feedback,
            
            # Outcome
            'rejection_reason': self.rejection_reason,
            'rejection_stage': self.rejection_stage,
            'withdrawal_reason': self.withdrawal_reason,
            
            # Placement
            'placement_start_date': self.placement_start_date.isoformat() if self.placement_start_date else None,
            'placement_end_date': self.placement_end_date.isoformat() if self.placement_end_date else None,
            'placement_duration_months': self.placement_duration_months,
            
            # Priority
            'priority': self.priority,
            'is_hot': self.is_hot,
            'follow_up_date': self.follow_up_date.isoformat() if self.follow_up_date else None,
            
            # Timestamps
            'submitted_at': self.submitted_at.isoformat() if self.submitted_at else None,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None,
        }
        
        # Rate info (optional, might be sensitive)
        if include_rates:
            result.update({
                'bill_rate': float(self.bill_rate) if self.bill_rate else None,
                'pay_rate': float(self.pay_rate) if self.pay_rate else None,
                'rate_type': self.rate_type,
                'currency': self.currency,
                'margin': self.margin,
                'margin_percentage': self.margin_percentage,
            })
        
        # Include related objects
        if include_candidate and self.candidate:
            result['candidate'] = {
                'id': self.candidate.id,
                'first_name': self.candidate.first_name,
                'last_name': self.candidate.last_name,
                'email': self.candidate.email,
                'phone': self.candidate.phone,
                'current_title': self.candidate.current_title,
                'skills': self.candidate.skills,
                'location': self.candidate.location,
                'visa_type': self.candidate.visa_type,
            }
        
        if include_job:
            if self.job_posting:
                result['job'] = {
                    'id': self.job_posting.id,
                    'title': self.job_posting.title,
                    'company': self.job_posting.company,
                    'location': self.job_posting.location,
                    'job_type': self.job_posting.job_type,
                    'is_remote': self.job_posting.is_remote,
                    'platform': self.job_posting.platform,
                    'job_url': self.job_posting.job_url,
                }
            elif self.is_external_job:
                # Return external job info when job_posting is null
                result['job'] = {
                    'id': None,
                    'title': self.external_job_title,
                    'company': self.external_job_company,
                    'location': self.external_job_location,
                    'job_type': None,
                    'is_remote': None,
                    'platform': 'external',
                    'job_url': self.external_job_url,
                }
        
        if include_submitted_by and self.submitted_by:
            result['submitted_by'] = {
                'id': self.submitted_by.id,
                'email': self.submitted_by.email,
                'first_name': self.submitted_by.first_name,
                'last_name': self.submitted_by.last_name,
            }
        
        if include_activities:
            activities_list = list(self.activities) if self.activities else []  # type: ignore
            result['activities'] = [
                activity.to_dict() for activity in activities_list[:10]  # Last 10 activities
            ]
        
        return result
    
    @property
    def margin(self) -> float | None:
        """Calculate margin (bill_rate - pay_rate)."""
        if self.bill_rate and self.pay_rate:
            return float(self.bill_rate - self.pay_rate)
        return None
    
    @property
    def margin_percentage(self) -> float | None:
        """Calculate margin percentage."""
        if self.bill_rate and self.pay_rate and self.bill_rate > 0:
            return round(float((self.bill_rate - self.pay_rate) / self.bill_rate * 100), 2)
        return None
    
    @property
    def days_since_submitted(self) -> int | None:
        """Calculate days since submission."""
        if self.submitted_at:
            return (datetime.utcnow() - self.submitted_at).days
        return None
    
    @property
    def is_active(self) -> bool:
        """Check if submission is still active."""
        return self.status in SubmissionStatus.active()
    
    @property
    def is_terminal(self) -> bool:
        """Check if submission has reached a terminal state."""
        return self.status in SubmissionStatus.terminal()
